src/gateway/sharding/shard_manager.rs

```diff
@@ -214,169 +195,177 @@ impl ShardManager {
+    /// for the given ID.
+    ///
+    /// If a shard has been queued but has not yet been initiated, then this will return `false`.
     #[must_use]
-    pub fn intents(&self) -> GatewayIntents {
-        self.gateway_intents
-    }
-
-    pub async fn return_with_value(&self, ret: Result<(), GatewayError>) {
-        if let Err(e) = self.return_value_tx.lock().await.send(ret).await {
-            tracing::warn!("failed to send return value: {}", e);
-        }
+    pub fn has(&self, shard_id: ShardId) -> bool {
+        self.runners.contains_key(&shard_id)
     }
 
-    pub fn shutdown_finished(&self, id: ShardId) {
-        if let Err(e) = self.shard_shutdown_send.unbounded_send(id) {
-            tracing::warn!("failed to notify about finished shutdown: {}", e);
-        }
+    /// Returns the [`ShardId`]s of the shards that have been instantiated and currently have a
+    /// valid [`ShardRunner`].
+    #[cfg_attr(feature = "tracing_instrument", instrument(skip(self)))]
+    #[must_use]
+    pub fn shards_instantiated(&self) -> Vec<ShardId> {
+        self.runners.keys().copied().collect()
     }
 
-    pub async fn restart_shard(&self, shard_id: ShardId) {
-        self.restart(shard_id).await;
-        if let Err(e) = self.shard_shutdown_send.unbounded_send(shard_id) {
-            tracing::warn!("failed to notify about finished shutdown: {e}");
-        }
+    /// Returns the [`ShardRunnerInfo`] corresponding to each running shard.
+    ///
+    /// Note that the shard runner also holds a copy of its info, which is why each entry is
+    /// wrapped in `Arc<Mutex<T>>`.
+    #[cfg_attr(feature = "tracing_instrument", instrument(skip(self)))]
+    #[must_use]
+    pub fn runner_info(&self) -> HashMap<ShardId, Arc<Mutex<ShardRunnerInfo>>> {
+        self.runners.iter().map(|(&id, (runner, _))| (id, Arc::clone(runner))).collect()
     }
 
-    pub async fn update_shard_latency_and_stage(
-        &self,
-        id: ShardId,
-        latency: Option<Duration>,
-        stage: ConnectionStage,
-    ) {
-        if let Some(runner) = self.runners.lock().await.get_mut(&id) {
-            runner.latency = latency;
-            runner.stage = stage;
-        }
+    /// Returns the gateway intents used for this gateway connection.
+    #[must_use]
+    pub fn intents(&self) -> GatewayIntents {
+        self.intents
     }
 }
 
 impl Drop for ShardManager {
     /// A custom drop implementation to clean up after the manager.
     ///
-    /// This shuts down all active [`ShardRunner`]s and attempts to tell the [`ShardQueuer`] to
-    /// shutdown.
+    /// This shuts down all active [`ShardRunner`]s.
     ///
     /// [`ShardRunner`]: super::ShardRunner
     fn drop(&mut self) {
-        drop(self.shard_queuer.unbounded_send(ShardQueuerMessage::Shutdown));
+        info!("Shutting down all shards");
+
+        for shard_id in self.shards_instantiated() {
+            self.shutdown(shard_id, 1000);
+        }
```

Comment:
Nit: This is making a Vec of ShardIds just to iterate through them. This could probably be improved by outlining the inner logic (well, just the warn tbh) of `ShardManager::shutdown` and using `drain` or `take`.

Reply:
I inlined the logic but I wonder if we should try waiting for shutdown to finish or not.

Reply:
We should follow File IO stuff, have a `close` or `shutdown_all` method which is async and has error handling but have a Drop impl which just spawns shutdown_all and hope it completes.

Reply:
Well, sending a shutdown message is sync already with `unbounded_send`, and the error handling is performed on the runner side, which will result in logging if errors occur. Is this sufficient you think?

Reply:
This is sufficent, a future PR could add a shutdown_all which waits for a response.